Fix server side treatment of '0', blank values, and spaces

apzeero opened the pull request:
This PR fixes the following:

- Submitted value of '0' is wrongly interpreted as empty
- '0' is saved and displayed as '(empty)'
- Leading and trailing spaces are not removed in submitted values

Renamed variable $isEmpty to $isBlank for readability. Simplified and minimized code changes as much as possible.

Related to #28

k1sul1 commented:
Hi, I don't know if you noticed but I fixed the "0" problem a few days ago? Because this seems duplicate. You've made your commits on the code I updated :) 

> Not sure about boolean `false` here. Is it intended for future use? Replaced with blank string for simplicity.

It's more explicit. I refuse to use empty strings to signify "no value". We have booleans and null, so I use them. 

Considering `$value = $formEntries[$name] ?? false;` equals `false` only when `$formEntries[$name]` is `NULL`, it wouldn't have made much sense to do `$formEntries[$name] ?? null;`, that would be redundant. 


> Leading and trailing spaces in submitted values can be problematic.
> Renamed variable `$isEmpty` to `$isBlank`, so less confusing with PHP `empty()` language construct. Used `$value === ''`, again for simplicity and readability.

We do not know what users want to do with their forms. Whitespace has to be left as is, as it would be a breaking change. Thus it is not possible to simply check for an empty string. Hence the eyesore I wrote.

If the user wants to ensure the submitted values are more than just blank spaces, then they add more spesific validation. The default validation is only meant to replicate the functionality of the browsers. You can pass a required field by entering just one space, unless more rules are present. 

I also do not find it possible to confuse a VARIABLE with a FUNCTION esp. when there are no function calls on the line.

apzeero commented:
> Hi, I don't know if you noticed but I fixed the "0" problem a few days ago? Because this seems duplicate. You've made your commits on the code I updated :)

Yes, saw your update. That's why I made this PR in draft mode for review and discussion.

> It's more explicit. I refuse to use empty strings to signify "no value". We have booleans and null, so I use them.

Fair enough.

> We do not know what users want to do with their forms. Whitespace has to be left as is, as it would be a breaking change. Thus it is not possible to simply check for an empty string. Hence the eyesore I wrote.

I wanted to propose a default behavior for this since it is the most common requirement I encounter. But I guess there is no need since the target users of libreform are developers, and they can easily remove spaces depending on their use case.

apzeero commented:
Closing this PR, but we still need to fix  `classes/modules/class-selectors.php` because '0' is still being saved and displayed as '(empty)'. I'll do a separate PR following your solution.

https://github.com/libreform/libreform/blob/c9a912a706bef5b4cc93e327848ba1938b042a17/classes/modules/class-selectors.php#L56

k1sul1 commented:
It would be possible to implement a feature flag that uses `wplfFieldsBeforeValidateSubmission` filter to mutate the values before validation and save. All it would need is a checkbox in the form edit view ("remove extra whitespace from entries?"), and if it's checked, use said filter and loop values. 

Just has to be off by default :)
